Enhancing lint.sh with additional checks for Chart.yaml

depohmel opened the pull request:
This PR is enhancing current CirckeCi initial linter.

It looks like most of the time submitted Chart.yaml (a super common case for new charts) has missing fields. this PR is enabling field validation against yaml-schemas/Chart.yaml schema file with https://github.com/23andMe/Yamale CLI tool.
Also, this change will allow adding schema[s] for values.yaml. 

Maintainer GitHub account validation is another new test. List of maintainers got parsed with yq tool and validated via curl against https://github.com/<account_name> 
this should make @unguiculus life slightly easy. 

Examples of outputs:
Successful validation
```
circleci@515ce1f45960:/charts$ sudo test/circle/lint.sh stable/kanister-operator/
From https://github.com/kubernetes/charts
 * branch            master     -> FETCH_HEAD
test/circle/lint.sh: line 78: Exit: command not found

Running helm dep build on the chart at stable/kanister-operator
No requirements found in stable/kanister-operator/charts.

Running helm lint on the chart at stable/kanister-operator
==> Linting stable/kanister-operator
Lint OK

1 chart(s) linted, no failures

Linting the Chart.yaml and values.yaml files at stable/kanister-operator
Validating /charts/stable/kanister-operator/Chart.yaml...
Validation success! 👍
Validating maintainers names

Checking the Chart version has increased for the chart at stable/kanister-operator
New higher version 0.2.1 found
```

Missing field in Chart.yaml
```
==> Linting stable/weave-cloud
Lint OK

1 chart(s) linted, no failures

Linting the Chart.yaml and values.yaml files at stable/weave-cloud
Validating /charts/stable/weave-cloud/Chart.yaml...

Error!
Schema: test/circle/yaml-schemas/Chart.yaml
Data file: /charts/stable/weave-cloud/Chart.yaml
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/yamale/command_line.py", line 29, in _validate
    yamale.validate(schema, data)
  File "/usr/local/lib/python3.6/site-packages/yamale/yamale.py", line 39, in validate
    schema.validate(d)
  File "/usr/local/lib/python3.6/site-packages/yamale/schema/schema.py", line 56, in validate
    raise ValueError(error_str)
ValueError: 
Error validating data /charts/stable/weave-cloud/Chart.yaml with schema test/circle/yaml-schemas/Chart.yaml
	appVersion: Required field missing

```

incorrect maintainer name
```
==> Linting stable/kanister-operator
Lint OK

1 chart(s) linted, no failures

Linting the Chart.yaml and values.yaml files at stable/kanister-operator
Validating /charts/stable/kanister-operator/Chart.yaml...
Validation success! 👍
Validating maintainers names
Error: depohmel65 is not valid GitHub account
```

Commit:
Enhancing lint.sh with additional checks

depohmel commented:
/assign @unguiculus 
/assign @mattfarina 
/assign @viglesiasce

viglesiasce commented on test/circle/lint.sh:
```diff
@@ -53,15 +53,29 @@ yamllinter() {
 source "$curDir/../semvercompare.sh"
-
-git remote add k8s https://github.com/kubernetes/charts
+#git remote add k8s https://github.com/kubernetes/charts
```
Was this removed intentionally?

depohmel replied:
I modified the file for local testing. this is post revert leftovers

viglesiasce commented on test/circle/lint.sh:
```diff
@@ -53,15 +53,29 @@ yamllinter() {
 CHANGED_FOLDERS=`git diff --find-renames --name-only $(git merge-base k8s/master HEAD) stable/ incubator/ | awk -F/ '{print $1"/"$2}' | uniq`
 
-# Exit early if no charts have changed
+Exit early if no charts have changed
```
This was uncommented for some reason

Commit:
reverting post local testing changes

mattfarina commented on test/circle/install.sh:
```diff
@@ -32,6 +32,14 @@ helm repo add incubator https://kubernetes-charts-incubator.storage.googleapis.c
+
+# Install a yq bash yaml cli tool
+# Pinning to a version for consistency
+sudo pip install yq==2.1.1
```
The tool installed below (`https://github.com/mikefarah/yaml`) is also a yaml query tool. It's been renamed yq (we're still using an older version). We should install one yaml querying tool and use that consistently.

I'm indifferent to the one we use.

mattfarina commented on test/circle/lint.sh:
```diff
@@ -53,15 +53,29 @@ yamllinter() {
+
+  for name in $(yq -r '.maintainers|.[]|.name' ${1}/Chart.yaml); do
+    if [ $(curl -s -o /dev/null -w "%{http_code}\n" -If https://github.com/${name}) -ne 200 ]; then
+      echo "Error: ${name} is not valid GitHub account"
```
I like this check. Would it be possible to add some polite language letting people know we have that as a requirement on the charts repo so that we can contact chart maintainers in issues/PRs.

mattfarina commented on test/circle/lint.sh:
```diff
@@ -53,15 +53,29 @@ yamllinter() {
 CHANGED_FOLDERS=`git diff --find-renames --name-only $(git merge-base k8s/master HEAD) stable/ incubator/ | awk -F/ '{print $1"/"$2}' | uniq`
 
-# Exit early if no charts have changed
+#Exit early if no charts have changed
```
The space before `Exit` was removed?

mattfarina commented on test/circle/yaml-schemas/Chart.yaml:
```diff
@@ -0,0 +1,13 @@
+---
+maintainer:
+  name: str()
+  email: str(required=False)
```
There are more than just these that can be in a chart. The options are described in [a proto file](https://github.com/kubernetes/helm/blob/master/_proto/hapi/chart/metadata.proto). Could we get more of the properties added to the schema?

mattfarina reviewed (commented):
In general, I like the additional checks and where this is going. Thanks for putting this together. I just have a little feedback.

Commit:
addressing reviews.

removing yq installetion
chaning yq with yaml for Chart.yaml parsing.
modifing Erro message for not github account check.

also modifing lint.sh so it can be used locally

depohmel commented:
I've modified lint.sh so it can be used locally. it can be removed and added as separate, better-structured PR. 
Once $1 arg is specified, all checks will be run against that folder.

I was trying to add something like lint_my_chart.sh, but got stuck with questions: where to place it, where should readme go, what execution to support etc
for now, it can be used with following CMD:
from charts folder

```
docker run -ti -v $(pwd .):/charts -w /charts/test circleci/python:3.6.4 bash -c "sudo circle/install.sh && circle/lint.sh ../stable/weave-cloud"
```

viglesiasce commented:
/approve

viglesiasce commented:
/ok-to-test

mattfarina commented:
/lgtm
